test/integration/xds_config_tracker_integration_test.cc

```diff
@@ -0,0 +1,256 @@
+                         DELTA_SOTW_UNIFIED_GRPC_CLIENT_INTEGRATION_PARAMS);
+
+TEST_P(XdsConfigTrackerIntegrationTest, XdsConfigTrackerSuccessCount) {
+  TestXdsConfigTrackerFactory factory;
+  Registry::InjectFactory<Config::XdsConfigTrackerFactory> registered(factory);
+
+  initialize();
+  acceptXdsConnection();
+
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Runtime, "", {"some_rtds_layer"},
+                                      {"some_rtds_layer"}, {}, true));
+  const auto some_rtds_layer = TestUtility::parseYaml<envoy::service::runtime::v3::Runtime>(R"EOF(
+    name: some_rtds_layer
+    layer:
+      foo: bar
+      baz: meh
+  )EOF");
+  sendDiscoveryResponse<envoy::service::runtime::v3::Runtime>(
+      Config::TypeUrl::get().Runtime, {some_rtds_layer}, {some_rtds_layer}, {}, "1");
+  test_server_->waitForCounterEq("test_xds_tracker.on_config_accepted", 1);
+  EXPECT_EQ(1, test_server_->counter("test_xds_tracker.on_config_accepted")->value());
+}
+
+TEST_P(XdsConfigTrackerIntegrationTest, XdsConfigTrackerFailureCount) {
+  TestXdsConfigTrackerFactory factory;
+  Registry::InjectFactory<Config::XdsConfigTrackerFactory> registered(factory);
+
+  initialize();
+  acceptXdsConnection();
+
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Runtime, "", {"some_rtds_layer"},
+                                      {"some_rtds_layer"}, {}, true));
+
+  const auto route_config =
+      TestUtility::parseYaml<envoy::config::route::v3::RouteConfiguration>(R"EOF(
+    name: my_route
+    vhds:
+      config_source:
+        resource_api_version: V3
+        api_config_source:
+          api_type: GRPC
+          transport_api_version: V3
+          grpc_services:
+            envoy_grpc:
+              cluster_name: xds_cluster
+    )EOF");
+
+  FakeStream* stream = xds_stream_.get();
+  if (sotw_or_delta_ == Grpc::SotwOrDelta::Sotw ||
+      sotw_or_delta_ == Grpc::SotwOrDelta::UnifiedSotw) {
+    envoy::service::discovery::v3::DiscoveryResponse discovery_response;
+    discovery_response.set_version_info("1");
+    discovery_response.set_type_url(Config::TypeUrl::get().Runtime);
+    discovery_response.add_resources()->PackFrom(route_config);
+
+    static int next_nonce_counter = 0;
+    discovery_response.set_nonce(absl::StrCat("nonce", next_nonce_counter++));
+
+    // Message's TypeUrl != Resource's
+    stream->sendGrpcMessage(discovery_response);
+  } else {
+    std::vector<envoy::service::discovery::v3::Resource> resources;
+    envoy::service::discovery::v3::Resource resource;
+    resource.mutable_resource()->PackFrom(route_config);
+    resource.set_name("my_route");
+    resource.set_version("1");
+    resources.emplace_back(resource);
+    const auto delta_discovery_response = createExplicitResourcesDeltaDiscoveryResponse(
+        Config::TypeUrl::get().Runtime, resources, {});
+
+    // Message's TypeUrl != Resource's
+    stream->sendGrpcMessage(delta_discovery_response);
+  }
```

Comment:
I'm probably missing something here, but why can't this just be a call to `sendDiscoveryResponse`. I'm thinking of something similar to the following:
```
sendDiscoveryResponse<envoy::service::runtime::v3::Runtime>(
      Config::TypeUrl::get().Runtime, {route_config}, {route_config}, {}, "1");
```
Will the above work instead of this code?

Reply:
This is a template method, and the type of resources must be same as the message's type_url as well as the template type. My code simulates a simple exception scenario which the types are not same.

Reply:
Thanks! `sendDiscoveryResponse<envoy::config::route::v3::RouteConfiguration>( Config::TypeUrl::get().Runtime, {route_config}, {route_config}, {}, "1");` works. I should know this.